Push new `demo` Docker image on PostRelease workflow

knelli2 opened the pull request:
## Proposed changes

Also add a new workflow so we can push to the `demo` image whenever we need to so we aren't restricted to just releases.

Also updates all the docs that were based on the old containers.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Build and push `demo` image in post release workflow

Also add a new workflow that allows us to push to the `demo` image
whenever we want so we aren't restricted to releases.

nilsvu commented on .github/workflows/Docker.yaml:
```diff
@@ -0,0 +1,31 @@
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3.0.1
+        with:
+          fetch-depth: 0
```
delete these lines (no need to fetch more than the latest commit)

nilsvu commented on .github/workflows/Docker.yaml:
```diff
@@ -0,0 +1,31 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+name: Docker
```
Maybe name this workflow (and the file) "Demo container"?

nilsvu commented on .github/workflows/Docker.yaml:
```diff
@@ -0,0 +1,31 @@
+name: Docker
+
+on:
+  workflow_dispatch:
```
You can make this workflow reusable. Just add the `workflow_call:` trigger here (https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow)

Then in the `PostRelease` workflow you can write:

```yaml
jobs:
  # ...
  docker:
    uses: ./.github/workflows/DemoContainer.yaml
```

So you don't have to duplicate the workflow.

nilsvu commented on docs/Installation/Installation.md:
````diff
@@ -149,55 +149,58 @@ To build with the Docker image:
 2. Retrieve the Docker image (you may need `sudo` in front of this command)
    ```
-   docker pull sxscollaboration/spectrebuildenv:latest
+   docker pull sxscollaboration/spectre:demo
````
We should use the `dev` container here, because the intention is to clone the spectre repo (often on the host file system), edit code, and build (in the container). The demo container doesn't work well for this, because it already has a cloned spectre repo (in the container). So please change `demo` -> `dev` here. You can probably move some or most of the other edits you made in this file to the "Hitchhiker's" guide. I'll review the changes to the guide then.

knelli2 committed:
Update docs to use new container

Updaes Installation, Hitchhiker's Guide, DockerVSCodeQuickStart,
and changes the cmake kit and devcontainer files.

knelli2 commented:
@nilsvu Thanks for all the useful tips! I squashed in all my changes directly. The `Installation.md` file still has a lot of changes, but they are mostly formatting changes. I only changed a few actual lines.

nilsvu commented on docs/Tutorials/BeginnersTutorial.md:
```diff
@@ -20,109 +20,103 @@ To start off, you'll need to obtain an environment to build and run SpECTRE in.
-dependencies pre-installed for you to use. Probably the easiest way to get
-started with this is to follow our \ref dev_guide_quick_start_docker_vscode
-tutorial which will have you
+dependencies pre-installed for you to use. The container also has the SpECTRE
+repository cloned in it already so you don't have to worry about getting it
+yourself. To obtain the docker image, run
```
How about "...has SpECTRE pre-installed so you don't have to worry about building it yourself"

nilsdeppe reviewed (approved):
Depending on timing, it might be nice to test this for the Sept. release :)

nilsvu reviewed (approved):
I'm happy with merging as-is in time for the Sep release and doing an edit pass over the beginner tutorial some time later, perhaps by a new person who worked through it and has some feedback.

knelli2 commented:
@nilsdeppe @nilsvu I think I'll leave the beginners guide as-is in the interest of getting this in before the September release to test things out.

I know @pajkosmi is just getting started with SpECTRE, so maybe if he goes through the beginners guide he can jot down some notes for improvement and open a follow up PR?
